eras/shelley/impl/src/Cardano/Ledger/Shelley/CertState.hs

```diff
@@ -0,0 +1,148 @@
+import Cardano.Ledger.Shelley.Era (ShelleyEra)
+import qualified Cardano.Ledger.UMap as UM
+import Control.DeepSeq (NFData (..))
+import Control.Monad.Trans (lift)
+import Data.Aeson (KeyValue, ToJSON (..), object, pairs, (.=))
+import Data.Default (Default (..))
+import qualified Data.Foldable as F
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+import Lens.Micro (Lens', lens, _2)
+import NoThunks.Class (NoThunks (..))
+
+data ShelleyCertState era = ShelleyCertState
+  { shelleyCertVState :: !(VState era)
+  , shelleyCertPState :: !(PState era)
+  , shelleyCertDState :: !(DState era)
+  }
+  deriving (Show, Eq, Generic)
+
+mkShelleyCertState :: VState era -> PState era -> DState era -> ShelleyCertState era
+mkShelleyCertState v p d =
+  ShelleyCertState
+    { shelleyCertVState = v
+    , shelleyCertPState = p
+    , shelleyCertDState = d
+    }
+
+shelleyCertDStateL :: Lens' (ShelleyCertState era) (DState era)
+shelleyCertDStateL = lens shelleyCertDState (\ds u -> ds {shelleyCertDState = u})
+{-# INLINE shelleyCertDStateL #-}
+
+shelleyCertPStateL :: Lens' (ShelleyCertState era) (PState era)
+shelleyCertPStateL = lens shelleyCertPState (\ds u -> ds {shelleyCertPState = u})
+{-# INLINE shelleyCertPStateL #-}
+
+shelleyCertVStateL :: Lens' (ShelleyCertState era) (VState era)
+shelleyCertVStateL = lens shelleyCertVState (\ds u -> ds {shelleyCertVState = u})
+{-# INLINE shelleyCertVStateL #-}
+
+toCertStatePairs :: KeyValue e a => ShelleyCertState era -> [a]
+toCertStatePairs ShelleyCertState {..} =
+  [ "dstate" .= shelleyCertDState
+  , "pstate" .= shelleyCertPState
+  ]
+
+shelleyObligationCertState :: ShelleyCertState era -> Obligations
+shelleyObligationCertState (ShelleyCertState VState {vsDReps} PState {psDeposits} DState {dsUnified}) =
+  let accum ans drepState = ans <> drepDeposit drepState
+   in Obligations
+        { oblStake = UM.fromCompact (UM.sumDepositUView (UM.RewDepUView dsUnified))
+        , oblPool = F.foldl' (<>) (Coin 0) psDeposits
+        , oblDRep = F.foldl' accum (Coin 0) vsDReps
+        , oblProposal = Coin 0
+        }
+
+shelleyCertsTotalDepositsTxBody ::
+  EraTxBody era => PParams era -> ShelleyCertState era -> TxBody era -> Coin
+shelleyCertsTotalDepositsTxBody pp ShelleyCertState {shelleyCertPState} =
+  getTotalDepositsTxBody pp (`Map.member` psStakePoolParams shelleyCertPState)
+
+shelleyCertsTotalRefundsTxBody ::
+  EraTxBody era => PParams era -> ShelleyCertState era -> TxBody era -> Coin
+shelleyCertsTotalRefundsTxBody pp ShelleyCertState {shelleyCertDState, shelleyCertVState} =
+  getTotalRefundsTxBody
+    pp
+    (lookupDepositDState shelleyCertDState)
+    (lookupDepositVState shelleyCertVState)
+
+instance EraCertState ShelleyEra where
+  type CertState ShelleyEra = ShelleyCertState ShelleyEra
+
+  mkCertState = mkShelleyCertState
+
+  upgradeCertState = error "It is not possible to translate Byron CertState with 'upgradeCertState'"
```

Comment:
It is good to add `Impossible`, cause this would result in a compile time error, since `ByronEra` does not have a `EraCertState` instance implemented
```suggestion
  upgradeCertState = error "Impossible: ByronEra does not have `EraCertState` instance"
```